pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -99,6 +113,20 @@ func NewCloudCIDRAllocator(client clientset.Interface, cloud cloudprovider.Inter
 		err := fmt.Errorf("cloudCIDRAllocator does not support %v provider", cloud.ProviderName())
 		return nil, err
 	}
+
+	stackType := stackIPv4
+
+	// Based on validation performed in startNodeIpamController(), if there are 2 service CIDRs provided,
+	// they are of different family types.
+
+	if isIP4(allocatorParams.ServiceCIDR) && isIP6(allocatorParams.SecondaryServiceCIDR) {
+		stackType = stackIPv4IPv6
+	} else if isIP6(allocatorParams.ServiceCIDR) && isIP4(allocatorParams.SecondaryServiceCIDR) {
+		stackType = stackIPv6IPv4
+	} else if isIP6(allocatorParams.ServiceCIDR) && !isIP4(allocatorParams.SecondaryServiceCIDR) {
+		stackType = stackIPv6
+	}
```

Comment:
the helpers function does not differentiate if there is a cidr or not, so an empty value is considered as false, and this makes this hard to read, as I assume that `!isIP6` means `isIPv4` so we must avoid using the negative on the conditions. This caused a considerable number of bugs in Kubernetes , we can make this easier to understand with the following approach https://go.dev/play/p/3fYZ_umOQiT

```
var stackType clusterStackType 
switch {
case isIP4(allocatorParams.ServiceCIDR) && allocatorParams.SecondaryServiceCIDR == nil:
   stackType = stackIPv4
case isIP6(allocatorParams.ServiceCIDR) && allocatorParams.SecondaryServiceCIDR == nil:
   stackType = stackIPv6
case isIP4(allocatorParams.ServiceCIDR) && allocatorParams.SecondaryServiceCIDR != nil:
   stackType = stackIPv4IPv6
case isIP6(allocatorParams.ServiceCIDR) && allocatorParams.SecondaryServiceCIDR != nil:
    stackType = stackIPv6IPv4
}

Reply:
Thanks! I changed the code like this:

```
         var stackType clusterStackType
 
         // Based on validation performed in startNodeIpamController(), if there are 2 service CIDRs provided,
         // they are of different family types.
 
         if isIP4(allocatorParams.ServiceCIDR) && isIP6(allocatorParams.SecondaryServiceCIDR) {
                 stackType = stackIPv4IPv6
         } else if isIP6(allocatorParams.ServiceCIDR) && isIP4(allocatorParams.SecondaryServiceCIDR) {
                 stackType = stackIPv6IPv4
         } else if isIP4(allocatorParams.ServiceCIDR) && allocatorParams.SecondaryServiceCIDR == nil {
                 stackType = stackIPv4
         } else if isIP6(allocatorParams.ServiceCIDR) && allocatorParams.SecondaryServiceCIDR == nil {
                 stackType = stackIPv6
         }
```

I like this form more, because it is more explicit. E. g.

```
if isIP4(allocatorParams.ServiceCIDR) && isIP6(allocatorParams.SecondaryServiceCIDR) {
                 stackType = stackIPv4IPv6
```

Means that both cidrs are non nil, the primary is ipv4 and the secondary is ipv6 and not representable as ipv4.

The condition is equivalent to what you proposed, but to understand well e. g.

```
case isIP4(allocatorParams.ServiceCIDR) && allocatorParams.SecondaryServiceCIDR != nil:
   stackType = stackIPv4IPv6
```

one needs to realize the primary cidr is non nil and ipv4, and the secondary cidr is non nil - so based on the validation performed earlier in `startNodeIpamController()`, the secondary cidr has to be ipv6. Because validation in this case is performed in two distant places, it may be harder to realize why the condition is correct. I know I added a comment explaining that, but still I think it might confuse someone.

I can also change to a switch if this helps (here probably the difference is only subtle).

WDYT? If you don't like my suggestion, I can just change to what you proposed.